### PR TITLE
Add load-time GPT-to-Hybrid checkpoint translation (#5675)

### PR DESCRIPTION
**Pull request by guihong-nv**
Allow HybridModel runs to load GPT (pure transformer) distributed checkpoints directly, without an offline conversion step. This supports both `torch_dist` and Megatron-FSDP (`fsdp_dtensor`) checkpoints, including model weights and distributed optimizer state. Instead of rewriting the checkpoint on disk, `load_checkpoint` retargets the hybrid model's own sharded state dict at the GPT checkpoint: attention and MLP entries are rewritten to the checkpoint's homogeneous layer format (layer index dropped from the storage key and prepended as a sharding axis), the final norm key is renamed, and layers without a GPT counterpart (for example Mamba) keep their fresh initialization via `LocalNonpersistentObject`. The regular distributed-checkpointing machinery then reads the GPT checkpoint directly and reshards across parallel-layout changes.

`--hybrid-layer-pattern` is required for GPT-to-Hybrid translation. `--finetune` is optional and keeps its normal meaning: a direct `--load` without it resumes iteration, optimizer, scheduler, RNG, and rerun state under normal compatibility rules; using `--finetune` resets training progress while still allowing translated optimizer state to load. `--no-load-optim` remains available for a weights-only load. The layer pattern must pair one-to-one with the checkpoint's layers (no MTP, no GDN/DSA positions, uniform dense or MoE MLP kind).

Also add `--use-legacy-model` so the hybrid entrypoint can keep training a GPT checkpoint as `GPTModel` without any translation.

Tests cover the pure layer-map/retarget logic and distributed model-plus-optimizer round trips across DP, Megatron FSDP, TP, PP, CP, EP, and ETP reconfigurations, including gated-MLP (SwiGLU factory) and grouped-GEMM MoE variants. Dedicated cases verify resume without `--finetune` for both `torch_dist` and `fsdp_dtensor` checkpoints.

This PR is required by #5675.
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by guihong-nv** [comment]
/ok to test [6cd111c](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/6cd111c62a14affe3dfd9120a063ce940e02fac1)

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -3331,6 +3331,13 @@ def _add_experimental_args(parser):
     group.add_argument('--hybrid-override-pattern', type=str, default=None,
                        help='Deprecated. Use --hybrid-layer-pattern instead. '
                        'If specified, its value will be forwarded to --hybrid-layer-pattern.')
+    group.add_argument('--use-legacy-model', action='store_true',
+                       help='Build a GPTModel from the hybrid training entrypoint '
+                       '(pretrain_hybrid.py) so an existing GPT checkpoint continues '
+                       'training as a GPT model instead of being translated to the '
+                       'hybrid layout. Incompatible with --hybrid-layer-pattern; '
+                       '--spec, if given, must be a GPT layer spec. Has no effect on '
+                       'other entrypoints.')
```
Comment (janEbert) [inline megatron/training/arguments.py]: I think we should discuss whether this should also apply for `pretrain_gpt.py`. The idea, after all, is that people get transferred to `HybridModel` automatically, even if they use deprecated code, unless they explicitly opt out.

**Comment by guihong-nv** [comment]
/ok to test [20ca49f](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/20ca49f68d75ba520a0f8b4fba488af18b094ba4)

**Comment by sbhavani** [comment]
Could we include a migration guide in markdown and a runtime warning? The runtime warning could clearly state that HybridModel is being automatically initialized and what carries over vs. reinitialized and how to opt out with legacy flag.

**Review by Phlip79 — APPROVED** [review]

File: docs/user-guide/hybrid-model-migration.md [hunk 2]
```diff
@@ -70,11 +70,88 @@ treated as a new architecture and benchmarked independently.
...
+#### Set the required flags
+
+When the hybrid run loads a GPT checkpoint, it must set:
+
+- `--hybrid-layer-pattern` — pairs the checkpoint's layers with hybrid layer
+  positions. Without it the load is rejected.
+- `--finetune` and `--no-load-optim` — hybrid layers without a GPT counterpart
+  keep their fresh initialization, and the GPT run's optimizer state and
+  iteration bookkeeping do not carry over. The load is rejected if either flag
+  is missing.
```
Comment (janEbert) [inline docs/user-guide/hybrid-model-migration.md]: I understand that you cannot load layers that don't exist, but we should be able to load and translate optimizer state, shouldn't we? I feel like these flags being required is a big limit for the general usefulness of the online conversion.

File: megatron/core/dist_checkpointing/gpt_checkpoint_interop.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    ShardedStateDict,
+    ShardedTensor,
+    ShardedTensorFactory,
+)
+from megatron.core.models.hybrid.hybrid_layer_allocation import (
+    Symbols,
+    get_layer_maps_from_layer_type_list,
+    parse_hybrid_pattern,
+)
+
```
Comment (jiemingz) [inline megatron/core/dist_checkpointing/gpt_checkpoint_interop.py:64]: is this the best place for this file? Could this go outside of `core` ? If not maybe in `megatron/core/dist_checkpointing`?
it feels a little unnatural because these changes don't deal with the definition of the hybrid model itself but rather a consequence of using a legacy feature with it

**Comment by guihong-nv** [comment]
/ok to test [8bb5377](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/8bb53774d1ac388f76ac375804d0e9a9d09c404b)

File: docs/user-guide/hybrid-model-migration.md (hunk 2 above)
Comment (guihong-nv) [reply docs/user-guide/hybrid-model-migration.md]: Fixed; pls take a look.

File: megatron/core/dist_checkpointing/gpt_checkpoint_interop.py (hunk 3 above)
Comment (guihong-nv) [reply megatron/core/dist_checkpointing/gpt_checkpoint_interop.py:64]: fixed

**Comment by guihong-nv** [comment]
/ok to test [9b78259](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/9b78259239fa991b7d76dc8c4057431fd09d7b31)

**Review by yaoyu-33 — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

**Comment by guihong-nv** [comment]
/ok to test [3c78038](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/3c780384c57c48b4bd642d549373706950e68739)

File: docs/user-guide/hybrid-model-migration.md (hunk 2 above)
Comment (janEbert) [reply docs/user-guide/hybrid-model-migration.md]: Thanks for allowing `load_optim=True`. Can't we also remove the `--finetune` flag requirement? We should be able to add logic to also work around the following, right?
> - `--finetune` — the hybrid architecture has a different layer count than the
>   GPT checkpoint, so iteration bookkeeping and the LR schedule restart fresh.
>   The load is rejected without it.

**Comment by guihong-nv** [comment]
/ok to test [0f73838](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/0f738386ccd9e04f27f0a8b3658203b4c27a0cb2)

**Comment by guihong-nv** [comment]
/ok to test [716875a](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/716875a8e20a9cb9d2873d73863fd05659b7e18d)

**Comment by guihong-nv** [comment]
/ok to test [49a44bd](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/49a44bde7faa43ce101d74cc2b8cee9143416983)

**Comment by guihong-nv** [comment]
/ok to test [49a44bd](https://github.com/NVIDIA/Megatron-LM/pull/5792/commits/49a44bde7faa43ce101d74cc2b8cee9143416983)
